internal/service/alarm_notification_searcher.go

```diff
@@ -0,0 +1,225 @@
+)
+
+type subscriptionInfo struct {
+	filters                search.Selector
+	uris                   string
+	consumerSubscriptionId string
+}
+
+// This file contains oran alarm notification serer search for matched subscriptions
+// at 1st step apply linear search
+type alarmSubscriptionSearcher struct {
+	logger *slog.Logger
+	jqTool *jq.Tool
+	//maps with prebuilt selector
+	subscriptionInfoMap *map[string]subscriptionInfo
+	pathIndexMap        *map[string]alarmSubIdSet
+	noFilterSubsSet     *alarmSubIdSet
+
+	//Parser used for the subscription filters
+	selectorParser *search.SelectorParser
+}
+
+func (b *alarmSubscriptionSearcher) SetLogger(
+	value *slog.Logger) *alarmSubscriptionSearcher {
+	b.logger = value
+	return b
+}
+
+func (b *alarmSubscriptionSearcher) SetJqTool(
+	value *jq.Tool) *alarmSubscriptionSearcher {
+	b.jqTool = value
+	return b
+}
+
+func (b *alarmSubscriptionSearcher) SetSelectorParser(
+	parser *search.SelectorParser) *alarmSubscriptionSearcher {
+	b.selectorParser = parser
+	return b
+}
+
+func newAlarmSubscriptionSearcher() *alarmSubscriptionSearcher {
+	return &alarmSubscriptionSearcher{}
+}
+
+func (b *alarmSubscriptionSearcher) build() {
+	b.subscriptionInfoMap = &map[string]subscriptionInfo{}
+	b.pathIndexMap = &map[string]alarmSubIdSet{}
+	b.noFilterSubsSet = &alarmSubIdSet{}
+
+	// Create the filter expression parser:
+	selectorParser, err := search.NewSelectorParser().
+		SetLogger(b.logger).
+		Build()
+	if err != nil {
+		b.logger.Error(
+			"failed to create filter expression parser: ",
+			slog.String("error: ", err.Error()),
+		)
+		return
+	}
+	b.selectorParser = selectorParser
+}
+
+// NOTE the function should be called by a function that is holding the semophone
+func (b *alarmSubscriptionSearcher) getSubFilters(filterStr string, subId string) (err error) {
+
+	//no filter found, return empty array and behavior as "*"
+	if filterStr == "" {
+		(*b.noFilterSubsSet)[subId] = struct{}{}
+		return
+	}
+
+	result, err := b.selectorParser.Parse(filterStr)
+
+	if err != nil {
+		return
+	}
+
+	subInfo := (*b.subscriptionInfoMap)[subId]
+	subInfo.filters = *result
+	(*b.subscriptionInfoMap)[subId] = subInfo
+
+	//for now use path 0 only
+	//to be fixed with full path for quicker search
+	for _, element := range result.Terms {
+		_, ok := (*b.pathIndexMap)[element.Path[0]]
+
+		if !ok {
+			(*b.pathIndexMap)[element.Path[0]] = alarmSubIdSet{}
+		}
+		(*b.pathIndexMap)[element.Path[0]][subId] = struct{}{}
+	}
+
+	return
+}
+
+// NOTE the function should be called by a function that is holding the semophone
+func (b *alarmSubscriptionSearcher) pocessSubscriptionMapForSearcher(subscriptionMap *map[string]data.Object,
+	jqTool *jq.Tool) (err error) {
+
+	for key, value := range *subscriptionMap {
+
+		(*b.subscriptionInfoMap)[key] = subscriptionInfo{}
+
+		subInfo := (*b.subscriptionInfoMap)[key]
+		//get uris
+		var uris string
+		err = jqTool.Evaluate(`.callback`, value, &uris)
```

Comment:
He have been using jq to extract fields from objects because till know we didn't need to know much about the fields and what they mean. But I'd say that is different for subscriptions: this code needs to know what "callback", "filter" and "consumerSubscriptionId" mean. So I'd say we should instead have a new `Subscription` type (probably in a new `models` package) and then use it instead of using jq.

Reply:
let's talk about this during code walkthrough